File: openmetadata-service/src/main/java/org/openmetadata/service/secrets/KubernetesSecretsManager.java
/*
 *  Copyright 2025 Collate
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.openmetadata.service.secrets;

import com.google.common.annotations.VisibleForTesting;
import io.kubernetes.client.openapi.ApiClient;
import io.kubernetes.client.openapi.ApiException;
import io.kubernetes.client.openapi.Configuration;
import io.kubernetes.client.openapi.apis.CoreV1Api;
import io.kubernetes.client.openapi.models.V1ObjectMeta;
import io.kubernetes.client.openapi.models.V1Secret;
import io.kubernetes.client.util.ClientBuilder;
import io.kubernetes.client.util.Config;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.openmetadata.schema.security.secrets.SecretsManagerProvider;
import org.openmetadata.service.exception.SecretsManagerException;
import org.openmetadata.service.exception.SecretsManagerUpdateException;

/**
 * Kubernetes implementation of the SecretsManager.
 * This implementation stores secrets as Kubernetes Secret objects.
 */
@Slf4j
public class KubernetesSecretsManager extends ExternalSecretsManager {
  private static final String NAMESPACE = "namespace";
  private static final String KUBECONFIG_PATH = "kubeconfigPath";
  private static final String IN_CLUSTER = "inCluster";
  private static final String SKIP_INIT = "skipInit";
  private static final String DEFAULT_NAMESPACE = "default";
  private static final String SECRET_KEY = "value";

  private static KubernetesSecretsManager instance = null;
  @Getter private CoreV1Api apiClient;
  private String namespace;

  private KubernetesSecretsManager(SecretsConfig secretsConfig) {
    super(SecretsManagerProvider.KUBERNETES, secretsConfig, 100);

    // Check if we should skip initialization (for testing)
    boolean skipInit =
        Boolean.parseBoolean(
            (String)
                secretsConfig
                    .parameters()
                    .getAdditionalProperties()
                    .getOrDefault(SKIP_INIT, "false"));

    if (!skipInit) {
      initializeKubernetesClient();
    }
  }

  public static KubernetesSecretsManager getInstance(SecretsConfig secretsConfig) {
    if (instance == null) {
      instance = new KubernetesSecretsManager(secretsConfig);
    }
    return instance;
  }

  private void initializeKubernetesClient() {
    try {
      ApiClient client;

      boolean inCluster =
          Boolean.parseBoolean(
              (String)
                  getSecretsConfig()
                      .parameters()
                      .getAdditionalProperties()
                      .getOrDefault(IN_CLUSTER, "false"));

      if (inCluster) {
        client = ClientBuilder.cluster().build();
        LOG.info("Using in-cluster Kubernetes configuration");
      } else {
        String kubeconfigPath =
            (String) getSecretsConfig().parameters().getAdditionalProperties().get(KUBECONFIG_PATH);
        if (StringUtils.isNotBlank(kubeconfigPath)) {
          client = Config.fromConfig(kubeconfigPath);
          LOG.info("Using kubeconfig from path: {}", kubeconfigPath);
        } else {
          // Default to ~/.kube/config
          client = Config.defaultClient();
          LOG.info("Using default kubeconfig");
        }
      }

      Configuration.setDefaultApiClient(client);
      this.apiClient = new CoreV1Api(client);

      // Set namespace
      this.namespace =
          (String)
              getSecretsConfig()
                  .parameters()
                  .getAdditionalProperties()
                  .getOrDefault(NAMESPACE, DEFAULT_NAMESPACE);
      LOG.info("Kubernetes SecretsManager initialized with namespace: {}", namespace);

    } catch (IOException e) {
      throw new SecretsManagerException("Failed to initialize Kubernetes client", e);
    }
  }

  @Override
  protected void storeSecret(String secretName, String secretValue) {
    String k8sSecretName = sanitizeSecretName(secretName);

    try {
      V1Secret secret = new V1Secret();
      V1ObjectMeta metadata = new V1ObjectMeta();
      metadata.setName(k8sSecretName);
      metadata.setNamespace(namespace);

      Map<String, String> labels = new HashMap<>();
      labels.put("app", "openmetadata");
      labels.put("managed-by", "openmetadata-secrets-manager");
      if (getSecretsConfig().tags() != null && !getSecretsConfig().tags().isEmpty()) {
        // Convert tags list to map if needed
        getSecretsConfig()
            .tags()
            .forEach(
                tag -> {
                  String[] parts = tag.split(":", 2);
                  if (parts.length == 2) {
                    labels.put(parts[0], parts[1]);
                  }
                });
      }
      metadata.setLabels(labels);
      secret.setMetadata(metadata);

      Map<String, byte[]> data = new HashMap<>();
      data.put(SECRET_KEY, secretValue.getBytes(StandardCharsets.UTF_8));
      secret.setData(data);

      apiClient.createNamespacedSecret(namespace, secret).execute();
      LOG.debug("Created Kubernetes secret: {}", k8sSecretName);

    } catch (ApiException e) {
      throw new SecretsManagerException(
          String.format("Failed to create Kubernetes secret: %s", k8sSecretName), e);
    }
  }

  @Override
  protected void updateSecret(String secretName, String secretValue) {
    String k8sSecretName = sanitizeSecretName(secretName);

    try {
      // Get existing secret
      V1Secret existingSecret = apiClient.readNamespacedSecret(k8sSecretName, namespace).execute();

      // Update the secret data
      Map<String, byte[]> data = new HashMap<>();
      data.put(SECRET_KEY, secretValue.getBytes(StandardCharsets.UTF_8));
      existingSecret.setData(data);

      apiClient.replaceNamespacedSecret(k8sSecretName, namespace, existingSecret).execute();
      LOG.debug("Updated Kubernetes secret: {}", k8sSecretName);

    } catch (ApiException e) {
      if (e.getCode() == 404) {
        // Secret doesn't exist, create it
        LOG.debug("Secret {} not found, creating new secret", k8sSecretName);
        storeSecret(secretName, secretValue);
      } else {
        throw new SecretsManagerUpdateException(
            String.format("Failed to update Kubernetes secret: %s", k8sSecretName), e);
      }
    }
  }

  @Override
  protected String getSecret(String secretName) {
    String k8sSecretName = sanitizeSecretName(secretName);

    try {
      V1Secret secret = apiClient.readNamespacedSecret(k8sSecretName, namespace).execute();

      if (secret.getData() != null && secret.getData().containsKey(SECRET_KEY)) {
        byte[] secretData = secret.getData().get(SECRET_KEY);
        return new String(secretData, StandardCharsets.UTF_8);
      }

      LOG.warn("Secret {} exists but has no data", k8sSecretName);
      return null;

    } catch (ApiException e) {
      if (e.getCode() == 404) {
        LOG.debug("Secret {} not found", k8sSecretName);
        return null;
      }
      throw new SecretsManagerException(
          String.format("Failed to retrieve Kubernetes secret: %s", k8sSecretName), e);
    }
  }

  @Override
  public boolean existSecret(String secretName) {
    String k8sSecretName = sanitizeSecretName(secretName);

    try {
      apiClient.readNamespacedSecret(k8sSecretName, namespace).execute();
      return true;
    } catch (ApiException e) {
      if (e.getCode() == 404) {
        return false;
      }
      throw new SecretsManagerException(
          String.format("Failed to check existence of Kubernetes secret: %s", k8sSecretName), e);
    }
  }

  @Override
  protected void deleteSecretInternal(String secretName) {
    String k8sSecretName = sanitizeSecretName(secretName);

    try {
      apiClient.deleteNamespacedSecret(k8sSecretName, namespace).execute();
      LOG.debug("Deleted Kubernetes secret: {}", k8sSecretName);
    } catch (ApiException e) {
      if (e.getCode() != 404) {
        throw new SecretsManagerException(
            String.format("Failed to delete Kubernetes secret: %s", k8sSecretName), e);
      }
      LOG.debug("Secret {} already deleted or does not exist", k8sSecretName);
    }
  }

  /**
   * Sanitize secret name to be Kubernetes compliant.
   * Kubernetes secret names must be lowercase alphanumeric or '-',
   * and must start and end with an alphanumeric character.
   */
  private String sanitizeSecretName(String secretName) {
    // Remove leading slashes
    String sanitized = secretName.replaceAll("^/+", "");
    sanitized = sanitized.replaceAll("[^a-zA-Z0-9-]", "-");
    sanitized = sanitized.toLowerCase();
    sanitized = sanitized.replaceAll("-+", "-");
    sanitized = sanitized.replaceAll("^-+|-+$", "");

    if (sanitized.length() > 253) {
      String hash = Integer.toHexString(secretName.hashCode());
      sanitized = sanitized.substring(0, 240) + "-" + hash;
    }
    if (!sanitized.matches("^[a-z0-9].*")) {
      sanitized = "om-" + sanitized;
    }
    return sanitized;
  }

  @VisibleForTesting
  void setApiClient(CoreV1Api apiClient) {
    this.apiClient = apiClient;
  }

  @VisibleForTesting
  void setNamespace(String namespace) {
    this.namespace = namespace;
  }
}

File: ingestion/src/metadata/utils/secrets/kubernetes_secrets_manager.py
#  Copyright 2025 Collate
#  Licensed under the Collate Community License, Version 1.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#  https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/LICENSE
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

"""
Kubernetes Secrets Manager implementation
"""
import base64
import os
import traceback
from abc import ABC
from typing import Optional

from kubernetes import client, config
from kubernetes.client.exceptions import ApiException

from metadata.generated.schema.security.credentials.kubernetesCredentials import (
    KubernetesCredentials,
)
from metadata.generated.schema.security.secrets.secretsManagerClientLoader import (
    SecretsManagerClientLoader,
)
from metadata.generated.schema.security.secrets.secretsManagerProvider import (
    SecretsManagerProvider,
)
from metadata.utils.dispatch import enum_register
from metadata.utils.logger import utils_logger
from metadata.utils.secrets.external_secrets_manager import (
    SECRET_MANAGER_AIRFLOW_CONF,
    ExternalSecretsManager,
    SecretsManagerConfigException,
)

logger = utils_logger()

secrets_manager_client_loader = enum_register()


def _get_current_namespace() -> str:
    """
    :return: The namespace where the application service account is running or default if it can't be retrieved
    """
    try:
        with open(
            "/var/run/secrets/kubernetes.io/serviceaccount/namespace", encoding="utf-8"
        ) as f:
            return f.read().strip()
    except Exception as _:
        logger.info(
            "Can't read the current namespace from in-cluster kubernetes. Is the service account configured?"
        )
    return "default"


# pylint: disable=import-outside-toplevel
@secrets_manager_client_loader.add(SecretsManagerClientLoader.noop.value)
def _() -> None:
    return None


@secrets_manager_client_loader.add(SecretsManagerClientLoader.airflow.value)
def _() -> Optional[KubernetesCredentials]:
    from airflow.configuration import conf

    namespace = conf.get(
        SECRET_MANAGER_AIRFLOW_CONF,
        "kubernetes_namespace",
        fallback=_get_current_namespace(),
    )
    in_cluster = conf.getboolean(
        SECRET_MANAGER_AIRFLOW_CONF, "kubernetes_in_cluster", fallback=False
    )
    kubeconfig_path = conf.get(
        SECRET_MANAGER_AIRFLOW_CONF, "kubernetes_kubeconfig_path", fallback=None
    )

    return KubernetesCredentials(
        namespace=namespace,
        inCluster=in_cluster,
        kubeconfigPath=kubeconfig_path,
    )


@secrets_manager_client_loader.add(SecretsManagerClientLoader.env.value)
def _() -> Optional[KubernetesCredentials]:
    namespace = os.getenv("KUBERNETES_NAMESPACE", _get_current_namespace())
    in_cluster = os.getenv("KUBERNETES_IN_CLUSTER", "false").lower() == "true"
    kubeconfig_path = os.getenv("KUBERNETES_KUBECONFIG_PATH")

    return KubernetesCredentials(
        namespace=namespace,
        inCluster=in_cluster,
        kubeconfigPath=kubeconfig_path,
    )


class KubernetesSecretsManager(ExternalSecretsManager, ABC):
    """
    Kubernetes Secrets Manager class
    """

    def __init__(
        self,
        loader: SecretsManagerClientLoader,
    ):
        super().__init__(provider=SecretsManagerProvider.kubernetes, loader=loader)

        # Initialize Kubernetes client
        if self.credentials.inCluster:
            config.load_incluster_config()
            logger.info("Using in-cluster Kubernetes configuration")
        else:
            kubeconfig_path = self.credentials.kubeconfigPath
            if kubeconfig_path:
                config.load_kube_config(config_file=kubeconfig_path)
                logger.info(f"Using kubeconfig from path: {kubeconfig_path}")
            else:
                config.load_kube_config()
                logger.info("Using default kubeconfig")

        self.client = client.CoreV1Api()
        self.namespace = self.credentials.namespace or _get_current_namespace()
        logger.info(
            f"Kubernetes SecretsManager initialized with namespace: {self.namespace}"
        )

    def get_string_value(self, secret_id: str) -> str:
        """
        :param secret_id: The secret id to retrieve
        :return: The value of the secret
        """
        try:

            # Get the secret from Kubernetes
            secret = self.client.read_namespaced_secret(
                name=secret_id, namespace=self.namespace
            )

            # Kubernetes stores secret data as base64 encoded
            if secret.data and "value" in secret.data:
                secret_value = base64.b64decode(secret.data["value"]).decode("utf-8")
                logger.debug(f"Got value for secret {secret_id}")
                return secret_value
            logger.warning(f"Secret {secret_id} exists but has no 'value' key")
            return None

        except ApiException as exc:
            if exc.status == 404:
                logger.debug(f"Secret {secret_id} not found")
                return None
            logger.debug(traceback.format_exc())
            logger.error(
                f"Could not get the secret value of {secret_id} due to [{exc}]"
            )
            raise exc
        except Exception as exc:
            logger.debug(traceback.format_exc())
            logger.error(
                f"Could not get the secret value of {secret_id} due to [{exc}]"
            )
            raise exc

    def load_credentials(self) -> Optional[dict]:
        """Load the provider credentials based on the loader type"""
        try:
            loader_fn = secrets_manager_client_loader.registry.get(self.loader.value)
            return loader_fn()
        except Exception as err:
            raise SecretsManagerConfigException(f"Error loading credentials - [{err}]")
